### DISCUSSION TITLE
대량 거래 내역 조회 시 JVM 튜닝을 통한 성능 최적화 전략: 메모리 관리, GC 튜닝, 힙 설정을 통한 효율적 데이터 처리 방안

### DISCUSSION THREAD

**Discussion by liebigbear** [100-hours-a-week/discussions#33] [Deepdive]
# JVM

### Java 프로그램뿐만 아니라 Java바이트코드로 컴파일된 다른 언어로 작성된 프로그램도 실행할 수 있게 만들어주는 가상머신

**JVM 사양**

- JVM(자바 가상 머신)은 사양에 의해 정의된 추상(가상) 컴퓨터
    - 추상컴퓨터 : 실제 하드웨어에 구성된 물리적인 컴퓨터가 아닌 소프트웨어로 구현된 개념적 컴퓨터
    - JVM은 특정 소프트웨어나 하드웨어에 종속되지 않도록 명확한 사양에 따라 정의된다.

**JVM 역할**

Java소스코드를 컴파일해서 생성된 바이트코드를 실행

바이트코드를 실행하기 위해 JVM다음 역할을 수행한다.

- Class Loader(클래스 로더) : 프로그램의 바이트 코드를 메모리에 로드합니다.
    - 클래스 로더 기본작업
        - 로드 : 유형의 바이너리 데이터를 찾아서 가져옴
        - 연결
            - 검증 : 가져온 유형의 정확성을 보장
            - 준비 : 클래스 변수를 위한 메모리를 할당하고 메모리를 기본값으로 초기화
            - 해결 : 유형의 기본 참조를 직접 참조로 변환
        - 초기화 : 클래스 변수를 시작값을 초기화 하는 Java코드를 호출
- 바이트 코드 해석 및 JIT 컴파일러 : 바이트코드를 운영 체제가 이해할 수 있는 기계어로 변환해 실행
- 메모리 관리 : Heap과 Stack같은 메모리 영역을 관리하며 GC를 통해 사용하지 않는 메모리 회수
- 스레드 관리 : 프로그램에서 사용하는 여러 스레드를 관리합니다.

# JVM Tuning

### 자바 프로그램에 실행될 때, 성능을 최적화 하기 위해 JVM의 설정을 조정하는 작업

**주요 튜닝 요소**

- 힙 메모리(Heap Memory) 설정
    - 힙 메모리는 JVM이 객체를 저장하는 데 사용하는 메모리 공간
    - JVM이 시작될 때 메모리 크기를 지정해줄 수 있다.
    - -Xms : JVM이 시작할 때 할당할 힙 메모리의 초기 크기를 지정한다.
    - -Xmx: JVM이 사용할 수 있는 최대 힙 메모리 크기를 지정한다.
- 가비지 컬렉션(Garbage Collection) 튜닝
    - 가비지 컬렉션은 더 이상 사용되지 않는 객체를 자동으로 정리하여 메모리를 회수하는 작업
    - 다양한 가비지 컬렉터가 있으며 특성에 맞게 선택할 수 있다.
    - G1
    - CMS
    - Paraller GC
- 스레드 스택 크기 설정
    - 스레드 스택 : 각 스레드가 사용하는 메모리 공간
    - -Xss : 스레드 스택 크기 조절
- JVM 로그 : -Xloggc를 통해 가비지 컬렉션 로그를 기록하여 성능 분석에 도움
- 클래스 데이터 공유(CDS) : 클래스 데이터를 미리 로드

**Comment by startupcodekr** [comment]
대량 거래 내역 조회 시 JVM 튜닝을 통한 성능 최적화 전략: 메모리 관리, GC 튜닝, 힙 설정을 통한 효율적 데이터 처리 방안
